rubix-core/src/main/java/com/qubole/rubix/core/CachingInputStream.java

```diff
@@ -329,24 +337,24 @@ else if (isCached.get(idx) == Location.CACHED) {
             chainedReadRequestChainBuilder.add(new ChainedReadRequestChain().addReadRequestChain(cachedReadRequestChain));
         }
 
-        if (nonLocalReadRequestChain != null ||
-                directReadRequestChain != null ||
+        if (directReadRequestChain != null ||
                 remoteReadRequestChain != null) {
             ChainedReadRequestChain shared = new ChainedReadRequestChain();
             if (remoteReadRequestChain != null) {
                 shared.addReadRequestChain(remoteReadRequestChain);
             }
 
-            if (nonLocalReadRequestChain != null) {
-                shared.addReadRequestChain(nonLocalReadRequestChain);
-            }
-
             if (directReadRequestChain != null) {
                 shared.addReadRequestChain(directReadRequestChain);
             }
             chainedReadRequestChainBuilder.add(shared);
         }
 
+        if (!nonLocalRequests.isEmpty()) {
+            for (NonLocalReadRequestChain nonLocalReadRequestChain1 : nonLocalRequests.values()) {
+                chainedReadRequestChainBuilder.add(new ChainedReadRequestChain().addReadRequestChain(nonLocalReadRequestChain1));
```

Comment:
Same like comment above, need not create new object of ChainedReadRequestChain just add nonLocalReadRequestChain1 to chainedReadRequestChainBuilder